text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+   */
+  requestResponse<T, U>(msg: IoRequest<T,U>): Promise<U | undefined>;
+}
+```
+
+For every message, the toolkit calls `notify` or `requestResponse`, respectively.
+
+A generic `IoEmitter` is available to provide a familiar event emitter interface.
+This implementation will do nothing for messages and requests that don't have a registered listener, i.e. the default response will be used as required.
+
+```ts
+const io = new IoEmitter();
+io.on('deploy', async (msg: IoMessage) => { /* do something */ });
+io.once('deploy:confirm-security', async (msg: IoRequest<boolean>): boolean => {
+  // reject the deployment
+  return false;
+})
+```
+
+The `CdkCommandLineIoHost` is available, implementing all the interaction behavior for the AWS CDK CLI [TODO: this might end up in a separate package]:
+
+```ts
+const io = new CdkCommandLineIoHost({
+  useTTY: true // optional, defaults to auto detecting if TTY should be used
+})
+```
+
+#### Cloud Executables
+
+A Cloud Executable is the primitive that defines a deployable AWS CDK application.
+AWS CDK apps might need to be synthesized multiple times with additional context values before they are ready.
+This concept is what the Cloud Executable represents.
+
+The most flexible way to obtain a Cloud Executable is from a function returning a `CloudAssembly`.
+Providing a function allows your implementation to integrate with the [Programmatic Toolkit]'s automatic context fetching system.
+The function will be called one or more times with the known context values of the current iteration.
+It must then use these values to synthesize and return a `CloudAssembly`.
+If the Cloud Assembly contains metadata that indicates that context information is missing, the toolkit will perform lookup AWS API calls using the appropriate credentials, and invoke the cloud executable again.
+For all features (e.g. lookups) to work correctly, the `App()` must be instantiated with the received context values.
+Since it is not possible to update the context of a `cdk.App`, it must be created inside the function scope.
+
+A basic implementation would look like this:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCloudAssemblyProducer(
+  async (context: Record<string, any>): cxapi.CloudAssembly => {
+    const app = new cdk.App({ context });
+    const stack = new cdk.Stack(app);
+    return app.synth();
+  }
+);
+```
+
+Existing applications, can also be used by simply passing an AWS CDK app directory (containing a `cdk.json` file):
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkApp("/path/to/cdk/app");
+```
+
+Alternatively the `cdk.out` directory from a previous synth can be used as well:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkOut("/path/to/cdk.out");
+```
+
+#### Actions - synth
+
+The `synth` action takes a Cloud Executable and synthesizes it in to a `CloudAssembly`.
+
+```ts
+const cx: ICloudExecutable =  await cdk.synth(cx, { stacks: ['MyTestStack'] });
```

Comment:
`CloudAssembly` is a data type, `ICloudExecutable` a behavioral type to create a `CloudAssembly`.

All CDK actions currently accept the behavioral type and implicitly create the assembly in the background. That makes `ICloudExecutable` somewhat a superset of `CloudAssembly` because you can always do something like

```ts
class StaticCloudExecutable implements ICloudExecutable {
  constructor(private readonly assmebly: CloudAssembly) {}

  public exec() {
    return this.assembly;
  }
}
```

Which seems simple enough.

If we don't do this, we have the options to:
- Make every action accept both `ICloudExecutable` and `CloudAssembly`
- Implement a dynamic `CloudAssembly` dynamic so that a call to `tryGetArtifact(id)` would first do what ever needs to be done to create the `CloudAssembly`.

I've opted for the most option that is imo the most explicit.

Reply:
I like what you did. The name is maybe just setting some expectations. `ICloudAssemblySource` ?

Reply:
Oh I do like `ICloudAssemblySource` 🤔